### ISSUE TITLE
Make sequence ID verification optional?

### ISSUE THREAD

**Issue by Whathecode** [cph-cachet/carp.core-kotlin#375]
Sequence IDs in data streams have a valid reason for being there:
- detecting dropped data
- allowing the backend to know no more data with older timestamps will be uploaded (unless sync points change, which is a separate topic)

However, to simplify client implementations, there has been a request to not have to be bothered with tracking sequence IDs. A specific infrastructure can therefore decide to omit sequence ID constraints. But, this inhibits using the current `DataStreamSequence` and `DataStreamBatch` objects, since these enforce these constraints.

Could there be an easy way for infrastructures to _disable_ sequence ID constraints? Is this at all desirable?

**Comment by bardram** [comment]
> Could there be an easy way for infrastructures to _disable_ sequence ID constraints? Is this at all desirable?

YES! This is highly desirable - this whole sequence ID issue is causing us many problems in CAWS right now.

**Comment by Whathecode** [comment]
Other than redeployment needing the last IDs, are you running into other issues?

**Comment by bardram** [comment]
See comments in #480

**Comment by Whathecode** [comment]
To [respond to comments on the duplicate issue](https://github.com/cph-cachet/carp.core-kotlin/issues/480#issuecomment-2901624825) (let's centralize design discussion here):

>  and we can't come up with any scenario where we don't just rely on the timestamps on the data.

As documented in the [CARP core paper](https://www.overleaf.com/project/6628a99ef3e55dc26a4091ce) (4.4.1 Data Reliability and Synchronization):

> Devices may have an internal clock which they use to timestamp the data they collect, or even only use a simple incrementing number assuming a fixed frequency at which data is collected. This clock may differ, as well as drift, from a global world clock, in particular for _connected_ devices which have no direct internet connectivity. It is the responsibility of the _client_ running on the _primary device_ to synchronize data to a global world clock.
This synchronization is represented by `SyncPoint` which contains the time of the global clock and the sensor timestamp at the point synchronization occurred, as well as the relative clock speed of the global clock compared to the sensor clock.
The default `SyncPoint.UnixEpoch` listed assumes that the sensor reports the number of microseconds since the Unix epoch, and that there is no clock drift (e.g., typical for internal sensors in smartphones).
> The timestamps in `Measurement` contain the _raw_ timestamps as reported by the sensor.
This fulfills the requirement of _lossless_ data collection, since the synchronization data is stored independently from raw timestamps.
Errors may occur during synchronization, or a `SyncPoint` may become outdated. E.g., clock speeds may decrease under lower voltages as a battery discharges.
To correct for this, other cues can be used in post-processing (e.g., correlations in accelerometer movements) to derive better synchronizations points, which would not be possible if only synchronized data is uploaded.
> Raw timestamps cannot be presumed to be sequential, e.g., they may start incrementing from zero each time the device is powered on, which can happen multiple times during a single study deployment.
Therefore, a `sequenceId` needs to be maintained by the client which is incremented for each collected data point.
In addition to determining data point order when re-synchronizing data, sequence IDs can also be used to detect out of order, or dropped, data on the data back-end.

**Comment by bardram** [comment]
We should (still) make this optional. Then, a `sequenceId` can (and not "needs") to be maintained by the client, which is incremented for each collected data point.

**Comment by Whathecode** [comment]
@yuanchen233 Is this the last plan for the 1.3 release? I didn't get around to looking into a design for this. Maybe you know CARP well enough by now to come up with a suggestion? :)

**Comment by yuanchen233** [comment]
> Is this the last plan for the 1.3 release?

Other than the bug fix in PR 536, this is the last issue we planned.

>  Is this at all desirable?

I think sequenceId currently carries too many responsibilities:

1. Identity: unique key for a point within a data stream.
2. Ordering: defines point order when timestamps are absent or unreliable.
3. Gap detection: missing IDs indicate potentially missing samples.
4. Append ordering: uploads are expected to arrive in increasing sequence order.
5. Watermark/finality: newer IDs implicitly mean older IDs will not arrive later.

The problem is that points 1-3 do not require upload order to be monotonic, but the current implementation enforces monotonic arrival because of points 4-5.

`sequenceId` is a central concept for `MutableDataStreamBatch`. If we want to make sequence IDs optional, we probably need a separate type, such as `UnsequencedDataStreamBatch`, rather than weakening the existing sequence model.

My preference is to keep `sequenceId`, but limit its responsibilities to points 1-3. Then we should introduce a separate concept for points 4-5, such as `receiveId` for backend ingestion order and/or `finalizedSequenceId` for explicit finality. This would allow the backend to accept valid non-overlapping sequence ranges even when they arrive out of order, without losing the benefits of sequence IDs.

> Could there be an easy way for infrastructures to disable sequence ID constraints?

Maybe we could introduce a `SequenceIdConstraintPolicy` class contains different policies, like `StrictAppendOrder`, `AllowOutOfOrder`. Then on infrastructure level, declare the policy needed when constructing the service.

**Comment by Whathecode** [comment]
Point 4 and 5 are only current constraints out of pragmatism. You can easily imagine supporting out of order delivery with a different criteria for when to consider data points as 'dropped', which holds non-sequential data in memory until preceding points have arrived. And, that would work equally well with the current sequenceID. I doubt an additional identifier is helpful here. 

Instead, I was thinking more along the lines of your thoughts on a configured policy (I would expect on a per deployment level, but maybe even per data stream). The DTO would then allow sequence ID to be nullable (or we could consider a magic value), and evaluate that based on the current policy: either require it, or don't allow it, and generate the ID on the back end.

Of course, I'd still strongly discourage using the latter policy, and for backwards compatibility it shouldn't be the default.

**Comment by yuanchen233** [comment]
That makes sense, the policy would then about: **who owns sequence ID assignment**, this also aligns with _"a sequenceId can (and not "needs") to be maintained by the client"_ from above.


> Of course, I'd still strongly discourage using the latter policy, and for backwards compatibility it shouldn't be the default.

Yes. With backend-assigned sequence IDs, we lose source data order and gap detection. This should be declared as not recommended.

For CAWS specifically, most `DataStreamPoint` have timestamps, and smartphones act as the collection middle layer, so sequence ID are less important as the primary ordering signal.

**Comment by Whathecode** [comment]
> For CAWS specifically, most `DataStreamPoint` have timestamps, and smartphones act as the collection middle layer, so sequence ID are less important as the primary ordering signal.

Unless you want to embed time zone information in those data points, as [was requested here](https://discord.com/channels/1470427188401541185/1470699101468360765/1494350860061245600). 😉 Then the need for sequenceId pops up again.

**Comment by yuanchen233** [comment]
I’m having second thoughts on this.

Going through the comments again, the original problem was lack of access to `lastSequenceId`, which has now been addressed by #527.

Other potential issues, such as out-of-order uploads, can be handled at the infrastructure level, as you mentioned earlier.

Speaking from the CAWS perspective: sequence IDs have not really been used in a way that lets us see their benefits yet. I would prefer to first see how the new endpoint from #527 works in practice.

Speaking from the Core perspective: I am not sure this is a desirable feature.
